### PR TITLE
VCB-14 Conditional jump

### PR DESCRIPTION
**Pull request by deflox**
@PhilSchni since you are marked as the reviewer for this story, be advised that this branch is based on the assembly branch from Claudio which is not yet merged into the master branch. So you will see a lot more changes that we actually did in this branch.

Our changes mainly consist of the new CMN, CMP and B[cond] instruction.

### CODE REVIEW

**Comment by deflox** [comment]
@PhilSchni since you are the reviewer for this PR, be advised that this branch is based on the assembly branch from Claudio which is not yet merged into the master branch. So you will see a lot more changes that we actually did in this branch.

Our changes mainly consist of the new CMN, CMP and B<cond> instruction.

**Comment by deflox** [comment]
@PhilSchni I now did the merge back from master, the changes are now only including the actual changes that were made in this story.

File: test/instruction/instructions/compare/cmp.test.ts [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+  it('should return correct result for CMP R3, #1; R3 = 0x80000000', () => {
+    let registerArray = ['R3', '#1']
+    let opcode = cmpImmInstruction.encodeInstruction(registerArray)
+    cmpImmInstruction.executeInstruction(opcode, registers, memory)
+    expect(registers.readRegister(Register.R3).value).toEqual(0x80000000)
+    expect(registers.isFlagSet(Flag.N)).toBeFalsy()
+    expect(registers.isFlagSet(Flag.Z)).toBeFalsy()
+    expect(registers.isFlagSet(Flag.C)).toBeTruthy()
+    expect(registers.isFlagSet(Flag.V)).toBeTruthy()
+  })
```
Comment (PhilSchni) [inline test/instruction/instructions/compare/cmp.test.ts:236]: I would recommend to also make tests with high values like 0x7fffffff similar like for cmn. Even if it is not a normal situation, we should check with the tests if the results will be as on the physical board.

File: test/instruction/instructions/jump/bconditional.test.ts [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+
+  test('test execution if flags are correctly set for BNE', () => {
+    reset()
+    registers.setFlag(Flag.Z, false)
+    execute(BNE, firstPCValue, secondPCValue)
+  })
+
+  test('test execution if flags are correctly set for BCS', () => {
+    reset()
+    registers.setFlag(Flag.C, true)
```
Comment (PhilSchni) [inline test/instruction/instructions/jump/bconditional.test.ts:315]: Should we check with all this tests if Flags which should not get set are left as is? (Maybe de function changes a wrong flag as well too.)

Comment (deflox) [reply test/instruction/instructions/jump/bconditional.test.ts:315]: I would argue against this. I mean this tests checks if the condition jumps if the flags are correctly set according to the table in our handout. If you check the instruction code there is never any call to `setFlag()` function on the `Register` class. So in my point of view a test like this is not really worth it.

File: test/instruction/instructions/compare/cmp.test.ts (hunk 1 above)
Comment (deflox) [reply test/instruction/instructions/compare/cmp.test.ts:236]: @taggatips this one is for you.

File: test/instruction/instructions/jump/bconditional.test.ts (hunk 2 above)
Comment (clooudch) [reply test/instruction/instructions/jump/bconditional.test.ts:315]: Given that these tests are unit tests and therefor white box tests, it's okay to have the knowledge in the test that there is no `setFlag` and the flags can never be changed.

File: test/instruction/instructions/compare/cmp.test.ts (hunk 1 above)
Comment (taggatips) [reply test/instruction/instructions/compare/cmp.test.ts:236]: What kind of comparison would you like to happen (second param)? From my point of view, the given tests account for most of the possible main results of subtraction ...

File: test/instruction/instructions/jump/bconditional.test.ts (hunk 2 above)
Comment (PhilSchni) [reply test/instruction/instructions/jump/bconditional.test.ts:315]: Okay, then we will leave it as it is currently.

**Review by PhilSchni — APPROVED** [review]

**Pull request merged**